crypto/kmac/kmac.c

```diff
@@ -0,0 +1,460 @@
+        return 1;
+    case EVP_MAC_CTRL_SET_SIZE:
+        size = va_arg(args, size_t);
+        kctx->out_len = size;
+        return 1;
+
+    case EVP_MAC_CTRL_SET_KEY:
+        p = va_arg(args, const unsigned char *);
+        len = va_arg(args, size_t);
+        if (len < 4 || len > KMAC_MAX_KEY) {
+            EVPerr(EVP_F_KMAC_CTRL, EVP_R_INVALID_KEY_LENGTH);
+            return 0;
+        }
+        /* The key encoding relies on md being set first */
+        if (EVP_MD_CTX_md(ctx) == NULL) {
+            EVPerr(EVP_F_KMAC_CTRL, EVP_R_NO_DIGEST_SET);
+            return 0;
+        }
+        return kmac_bytepad_encode_key(kctx->key, &kctx->key_len, p, len,
+                                       EVP_MD_CTX_block_size(ctx));
+
+    case EVP_MAC_CTRL_SET_CUSTOM:
+        p = va_arg(args, const unsigned char *);
+        len = va_arg(args, size_t);
+        if (len > KMAC_MAX_CUSTOM) {
+            EVPerr(EVP_F_KMAC_CTRL, EVP_R_INVALID_CUSTOM_LENGTH);
+            return 0;
+        }
+        return encode_string(kctx->custom, &kctx->custom_len, p, len);
+    default:
+        return -2;
+    }
+}
+
+static int kmac_ctrl_int(EVP_MAC_IMPL *kctx, int cmd, ...)
+{
+    int rv;
+    va_list args;
+
+    va_start(args, cmd);
+    rv = kmac_ctrl(kctx, cmd, args);
+    va_end(args);
+
+    return rv;
+}
+
+static int kmac_ctrl_str_cb(void *kctx, int cmd, void *buf, size_t buflen)
+{
+    return kmac_ctrl_int(kctx, cmd, buf, buflen);
+}
+
+static int kmac_ctrl_str(EVP_MAC_IMPL *kctx, const char *type,
+                         const char *value)
+{
+    if (value == NULL)
+        return 0;
+
+    if (strcmp(type, "digest") == 0) {
+        const EVP_MD *md = EVP_get_digestbyname(value);
+        if (md == NULL)
+            return 0;
+        return kmac_ctrl_int(kctx, EVP_MAC_CTRL_SET_MD, md);
+    }
+    if (strcmp(type, "outlen") == 0)
+        return kmac_ctrl_int(kctx, EVP_MAC_CTRL_SET_SIZE, atoi(value));
+    if (strcmp(type, "xof") == 0)
+        return kmac_ctrl_int(kctx, EVP_MAC_CTRL_SET_FLAGS, atoi(value));
+    if (strcmp(type, "key") == 0)
+        return EVP_str2ctrl(kmac_ctrl_str_cb, kctx, EVP_MAC_CTRL_SET_KEY,
+                            value);
+    if (strcmp(type, "hexkey") == 0)
+        return EVP_hex2ctrl(kmac_ctrl_str_cb, kctx, EVP_MAC_CTRL_SET_KEY,
+                            value);
+    if (strcmp(type, "custom") == 0)
+        return EVP_str2ctrl(kmac_ctrl_str_cb, kctx, EVP_MAC_CTRL_SET_CUSTOM,
+                            value);
+    if (strcmp(type, "hexcustom") == 0)
+        return EVP_hex2ctrl(kmac_ctrl_str_cb, kctx, EVP_MAC_CTRL_SET_CUSTOM,
+                            value);
+    return -2;
+}
+
+/*
+ * Encoding/ Padding Methods.
+ */
+
+/* Returns the number of bytes required to store 'bits' into a byte array */
+static unsigned int get_encode_size(size_t bits)
+{
+    unsigned int cnt, sz = sizeof(size_t);
+
+    for (cnt = 0; bits && (cnt < sz); ++cnt, bits >>= 8) {
```

Comment:
What is the style guide for an empty body loop?

Reply:
Perhaps `continue;` in the body would be better?

Reply:
... Or rewrite it as a while loop.

Reply:
I like continue, as Matt said.  Empty braces make me thing something was left out.

Reply:
I know that the single empty statement (i.e. the `;`) has been met with protests before...  but hey, opinions change over time...

Reply:
390 occurences of ```continue;``` and I bet most of them are for empty-body loops.

Reply:
No they are not :)